pip_review/__main__.py

```diff
@@ -211,6 +217,19 @@ def get_outdated_packages(forwarded):
         return packages
 
 
+def get_values_maxlength(outdated):
+    temp = (sub['name'] for sub in outdated)
+    name = max(len(elem) for elem in temp if elem is not None)
+    name = max([name, 7])
+    temp = (sub['version'] for sub in outdated)
+    version = max(len(elem) for elem in temp if elem is not None)
+    version = max([version, 7])
+    temp = (sub['latest_version'] for sub in outdated)
+    latest = max(len(elem) for elem in temp if elem is not None)
+    latest = max([latest, 6])
```

Comment:
Along with the comment above, I'm sorry for that.
Now that I have come up with a more concise process, I will replace it with the following.
```
    namelen = max(len('Package'), *(len(e['name']) for e in outdated))
    versionlen = max(len('Version'), *(len(e['version']) for e in outdated))
    latestlen = max(len('Latest'), *(len(e['latest_version']) for e in outdated))
```

Reply:
That is *still* repeating yourself! Also, it is still mixing the concern of extracting the data from `outdated` with the concern of formatting a table.

Reply:
I don't understand what the point was.

What I'm doing here is finding the maximum length of the string that each column contains.
Depending on the contents of the outdated packages, they have different lengths and cannot be processed with just a predetermined length.

It looks like I'm extracting the data here because otherwise I can't get the string length of each row in the specified column.
Using the maximum string length of each column obtained in this way, the layout of the table is created so that it looks neat.

After that, a table is created based on the data of each package.

Is it better to judge the length at the same time as making the table?
I feel that such processing is more complicated and easier for bugs to enter.

Reply:
Look at this.

```py
from operator import itemgetter

# Single source of truth about columns. The remainder of the code
# makes no assumptions about names, order or number of columns.
COLUMNS = {
    'Package': 'name',
    'Version': 'version',
    'Latest': 'latest_version',
    'Type': 'latest_filetype',
}

# Next two functions describe how to collect data for the
# table. Note how they are not concerned with columns widths.

def extract_column(data, field, title):
    return [title] + list(map(itemgetter(field), data))

def extract_table(outdated):
    return [
        extract_column(outdated, field, title)
        for title, field in COLUMNS.items()
    ]

# Next two functions describe how to format any table. Note that
# they make no assumptions about where the data come from.

def column_width(column):
    return max(map(len, filter(None, column)))

def format_table(columns):
    widths = list(map(column_width, columns))
    row_fmt = ' '.join(map('{{:<{}}}'.format, widths)).format
    ruler = '-' * (sum(widths) + len(widths) - 1)
    rows = list(map(row_fmt, *columns))
    head = rows[0]
    body = rows[1:]
    return '\n'.join([head, ruler] + body + [ruler])

# Clean separation of concerns, no repetition!

def main():
    # ...
    logger.info(format_table(extract_table(outdated)))
```